### ISSUE TITLE
Lab 5.2 Test Case 9_15

### ISSUE THREAD

**Issue by bevteo** [nus-cs2030/2223-s2#707]
Hi has anyone encountered a time limit exceeded problem for test case 9_15? Currently, my wait event immediately generates the check event and in check event is where I will check if the server that the customer is waiting at is available to serve the customer. If the customer is waiting at the SelfServer queue, and cannot be served yet, a check event with the next available SelfServer timestamp is generated.  I've had no troubles with the infinite loop problem up till this test case and I'm not sure what else could go wrong?

**Comment by Aron113** [comment]
you should check that the Check event is returned with the next free self check time. Might be helpful to use some print statements to check that this is happening

**Comment by gelalallaal** [comment]
i think might be something with the selfserver, self server test cases are only added from test case 9.

**Comment by bevteo** [comment]
Hi, i've checked my Check event and everything seems fine, I'm returning the self server's nextavail time and my check event always seems to terminate. Could there be other causes of an infinite loop somewhere else?

**Comment by sp4ce-cowboy** [comment]
> Hi, i've checked my Check event and everything seems fine, I'm returning the self server's nextavail time and my check event always seems to terminate. Could there be other causes of an infinite loop somewhere else?

What events do your check event generate? It could be useful to check all events that return themselves, usually infinite loops occur either because of faulty queuing mechanism (where the same customer is repeatedly added to the same queue) or because of the same event being added back to the PQ (events that return themselves)

**Comment by bevteo** [comment]
> > Hi, i've checked my Check event and everything seems fine, I'm returning the self server's nextavail time and my check event always seems to terminate. Could there be other causes of an infinite loop somewhere else?
> 
> What events do your check event generate? It could be useful to check all events that return themselves, usually infinite loops occur either because of faulty queuing mechanism (where the same customer is repeatedly added to the same queue) or because of the same event being added back to the PQ (events that return themselves)

Hi, my Check event currently only returns either another Check event or a Serve event. As for the Check event, I've checked and ensured that the new Check Event generated always has a different later timestamp. For the queuing mechanism that you mentioned, does it mean that my SelfServers are not allowed to have their own queues? Currently, my SelfServers all have their own queue and I basically just add the customer to all the SelfServer queues at once and when a customer is done being served the customer is being removed from all the SelfServer queues.

**Comment by sp4ce-cowboy** [comment]
@bevteo Yup that's fine, I had the same implementation. You just have to be careful when you iterate through and add to all self-checkouts, and remove from all self-checkouts. There's quite a few errors that can occur during the iteration process, maybe you can try checking all your list iterations again.
